File: src/sudoku_solver.cpp
#include <iostream>
#include <string>
#include <vector>
#include <string_view>
#include <array>
using namespace std;
constexpr string_view str { ".123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz<=>" };
constexpr size_t str_size = str.size();
constexpr auto arr = []() -> array<size_t, 128> {
    array<size_t, 128> a {0};
    for(size_t i = 0; i < str_size; ++i) {
        a.at(str[i]) = i;
    }
    return a;
}();
void print_grid(size_t& grid_size, vector<vector<size_t>>& grid) {
    for(size_t i = 0; i < grid_size; ++i) {
        for(size_t j = 0; j < grid_size; ++j) {
            cout << str[grid[i][j]];
        }
        cout << '\n';
    }
}
bool solve_sudoku(size_t& rows, size_t& cols, size_t& grid_size,
                  vector<vector<size_t>>& grid,
                  vector<vector<bool>>& row_contains,
                  vector<vector<bool>>& col_contains,
                  vector<vector<bool>>& subgrid_contains,
                  vector<pair<size_t, size_t>>& empty_cells) {
    bool backtrack = false;
    size_t placed_cells = 0;
    size_t n, S;
    size_t empty_cells_size = empty_cells.size();
    while(true) {
        if(placed_cells == empty_cells_size)
            return true;
        auto& [i, j] = empty_cells[placed_cells];
        S = (i / rows) * rows + j / cols;
        n = grid[i][j];
        if(backtrack) {
            grid[i][j] = 0;
            backtrack = false;
            row_contains[i][n-1] = false;
            col_contains[j][n-1] = false;
            subgrid_contains[S][n-1] = false;
        }
        for(; n < grid_size; ++n) {
            auto r = row_contains[i][n];
            auto c = col_contains[j][n];
            auto s = subgrid_contains[S][n];
            if(!(r || c || s)) {
                grid[i][j] = n + 1;
                ++placed_cells;
                r = true;
                c = true;
                s = true;
                break;
            }
        }
        if(n == grid_size) {
            if(!placed_cells)
                return false;
            --placed_cells;
            backtrack = true;
        }
    }
}
int main() {
    size_t rows, cols; // subgrid dimensions
    size_t grid_size; // square grid
    cout << "Enter the subgrid dimensions (rows by columns): ";
    cin >> rows >> cols;
    grid_size = rows * cols;
    if(!(grid_size && grid_size < str_size)) {
        cerr << "Error: Grid size can't be ";
        if(grid_size)
            cerr << "larger than ";
        cerr << (grid_size ? str_size - 1 : 0) << "\nNote: Grid size is determined by the number of cells in subgrid.\n";
        return 1;
    }
    vector<vector<size_t>> grid(grid_size, vector<size_t>(grid_size));
    vector<vector<bool>> row_contains(grid_size, vector<bool>(grid_size, false));
    vector<vector<bool>> col_contains(grid_size, vector<bool>(grid_size, false));
    vector<vector<bool>> subgrid_contains(grid_size, vector<bool>(grid_size, false));
    vector<pair<size_t, size_t>> empty_cells;
    empty_cells.reserve(grid_size);
    cout << "Enter the " << grid_size << " x " << grid_size << " sudoku grid row by row.\n"
            "  - Use 1-9 for a number 1-9 cell.\n"
            "  - Use A-Z for a number 10-35 cell.\n"
            "  - Use a-z for a number 36-61 cell.\n"
            "  - Use <=> for a number 62-64 cell.\n"
            "Any other character will be interpreted as an empty cell.\n";
    for(size_t i = 0, S; i < grid_size; ++i) {
        string row_str;
        cin >> row_str;
        if(row_str.size() < grid_size) {
            cerr << "Error: Insufficient input!\n";
            return 1;
        }
        for(size_t j = 0; j < grid_size; ++j) {
            size_t& k = grid[i][j];
            if((k = arr[row_str[j]]) && (k <= grid_size || (k = 0))) {
                S = (i / rows) * rows + j / cols;
                auto r = row_contains[i][k-1];
                auto c = col_contains[j][k-1];
                auto s = subgrid_contains[S][k-1];
                if(r || c || s) {
                    cerr << "Error: Invalid sudoku grid!\n";
                    return 1;
                }
                r = true;
                c = true;
                s = true;
            } else
                empty_cells.emplace_back(i, j);
        }
    }
    cout << "Solving...\n";
    if(solve_sudoku(rows, cols, grid_size, grid, row_contains, col_contains, subgrid_contains, empty_cells))
        cout << "Solution found:\n";
    else
        cout << "No solution exists for the sudoku grid!\n";
    print_grid(grid_size, grid);
}

File: src/sudoku_solver.rs
use mimalloc::MiMalloc;
#[global_allocator]
static GLOBAL: MiMalloc = MiMalloc;
use tracing::*;
use scan_rules::*;
use tracing_subscriber::EnvFilter;
use tracing_subscriber::fmt::time::Uptime;
use tracing_subscriber::fmt::format::FmtSpan;
use ndarray::prelude::*;
use std::io::{BufRead as _, stdin, IsTerminal as _};
use core::char::from_u32;
#[expect(clippy::as_conversions, reason = "should be safe to convert char to usize directly")]
fn print_grid(grid: &Array2<usize>) {
    let (r, c) = grid.dim();
    for i in 0..r {
        for j in 0..c {
            let k = match grid[[i, j]] {
                n @ 1..=9 => n + '0' as usize,
                n @ 10..=35 => n + 'A' as usize - 10,
                n @ 36..=61 => n + 'a' as usize - 36,
                n @ 62..=64 => n + '<' as usize - 62,
                _ => '.' as usize,
            };
            print!("{}", from_u32(k as u32).unwrap());
        }
        println!();
    }
}
#[instrument(skip_all)]
#[expect(nonstandard_style, reason = "temporary variable names")]
fn solve_sudoku(rows: usize, cols: usize, grid: &mut Array2<usize>, row_contains: &mut Array2<bool>, col_contains: &mut Array2<bool>, subgrid_contains: &mut Array2<bool>, empty_cells_len: usize) -> bool {
    let grid_size = rows * cols;
    let mut placed_cells = 0_usize;
    let mut backtrack = false;
    let mut empty_cells: Vec<(usize, usize)> = Vec::with_capacity(empty_cells_len);
    for ((i, j), k) in grid.indexed_iter() {
        if *k == 0 {
            empty_cells.push((i, j));
        }
    }
    'cell: loop {
        if placed_cells == empty_cells_len {
            return true;
        }
        let (i, j) = empty_cells[placed_cells];
        let S = (i / rows) * rows + j / cols;
        let mut n = grid[[i, j]];
        if backtrack {
            backtrack = false;
            grid[[i, j]] = 0;
            row_contains[[i, n-1]] = false;
            col_contains[[j, n-1]] = false;
            subgrid_contains[[S, n-1]] = false;
        }
        while n < grid_size {
            let r = &mut row_contains[[i, n]];
            let c = &mut col_contains[[j, n]];
            let s = &mut subgrid_contains[[S, n]];
            if !(*r || *c || *s) {
                grid[[i, j]] = n + 1;
                placed_cells += 1;
                *r = true;
                *c = true;
                *s = true;
                continue 'cell;
            }
            n += 1;
        }
        if placed_cells == 0 {
            return false;
        }
        placed_cells -= 1;
        backtrack = true;
    }
}
fn main() {
    let filter = EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info"));
    tracing_subscriber::fmt()
        .with_env_filter(filter)
        .with_target(false)
        .with_line_number(true)
        .with_span_events(FmtSpan::NEW | FmtSpan::CLOSE)
        .with_timer(Uptime::default())
        .init();
    let _main_span = info_span!("main").entered();
    let term = stdin().is_terminal();
    info!(terminal = ?term);
    let (rows, cols, grid_size) = loop {
        if term {
            print!("Enter the subgrid dimensions (rows by columns): ");
        }
        let result = try_readln! {
            (let rows: usize, let cols: usize) => (rows, cols)
        };
        match result {
            Ok((rows, cols)) => {
                let grid_size = rows * cols;
                if grid_size == 0 {
                    error!(?rows, ?cols, "Invalid input");
                } else if grid_size > 64 {
                    error!(?rows, ?cols, ?grid_size, "Grid size can't be larger than 64");
                } else {
                    break (rows, cols, grid_size);
                }
                if !term {return;}
            },
            Err(e) => {
                error!(error = %e, "Invalid input");
                if !term {return;}
            },
        }
    };
    let mut empty_cells_len = grid_size * grid_size;
    debug!(?rows, ?cols, ?grid_size, ?empty_cells_len);
    let mut grid: Array2<usize> = Array::zeros((grid_size, grid_size));
    let mut row_contains: Array2<bool> = Array::from_elem((grid_size, grid_size), false);
    let mut col_contains: Array2<bool> = Array::from_elem((grid_size, grid_size), false);
    let mut subgrid_contains: Array2<bool> = Array::from_elem((grid_size, grid_size), false);
    if term {
        println!(r"Enter the {grid_size}x{grid_size} sudoku grid row by row.
  - Use 1-9 for a number 1-9 cell.
  - Use A-Z for a number 10-35 cell.
  - Use a-z for a number 36-61 cell.
  - Use <=> for a number 62-64 cell.
Any other character will be interpreted as an empty cell.
Row string input with insufficient length will leave the remaining cells empty.");
    }
    {
        let mut iterator = stdin().lock().lines();
        for i in 0..grid_size {
            let row_str = iterator.next().unwrap().unwrap();
            for (j, c) in row_str.chars().enumerate() {
                if j == grid_size {
                    break;
                }
                #[expect(clippy::as_conversions, reason = "should be safe to convert char to usize directly")]
                let k: usize = match c {
                    '1'..='9' => c as usize - '0' as usize,
                    'A'..='Z' => c as usize - 'A' as usize + 10,
                    'a'..='z' => c as usize - 'a' as usize + 36,
                    '<'..='>' => c as usize - '<' as usize + 62,
                    _ => 0,
                };
                #[expect(nonstandard_style, reason = "temporary variable names")]
                if k != 0 && k <= grid_size {
                    let S = (i / rows) * rows + j / cols;
                    let r = &mut row_contains[[i, k-1]];
                    let C = &mut col_contains[[j, k-1]];
                    let s = &mut subgrid_contains[[S, k-1]];
                    if *r || *C || *s {
                        error!(ch = ?c, n = ?k, r = ?i, c = ?j, "Invalid sudoku grid!");
                        return;
                    }
                    empty_cells_len -= 1;
                    grid[[i, j]] = k;
                    *r = true;
                    *C = true;
                    *s = true;
                }
            }
        }
    }
    debug!(?empty_cells_len);
    let solvable = solve_sudoku(rows, cols, &mut grid, &mut row_contains, &mut col_contains, &mut subgrid_contains, empty_cells_len);
    if solvable {
        info!("Solution found:");
    } else {
        warn!("Unsolvable!");
    }
    print_grid(&grid);
}
